In [1]:
import numpy as np

### Part 1 — Understand the Array

In [2]:
results = np.array([
    [0.82, 0.78, 0.80],
    [0.91, 0.88, 0.90],
    [0.76, 0.81, 0.79],
    [0.88, 0.84, 0.86],
    [0.69, 0.72, 0.70]
])

##### Q1

In [3]:
print(results)

[[0.82 0.78 0.8 ]
 [0.91 0.88 0.9 ]
 [0.76 0.81 0.79]
 [0.88 0.84 0.86]
 [0.69 0.72 0.7 ]]


In [4]:
print("Array Properties:")
print(f"Shape: {results.shape}") 
print(f"No of Dimentions: {results.ndim}")
print(f"Size: {results.size}")
print(f"Data Type: {results.dtype}")

Array Properties:
Shape: (5, 3)
No of Dimentions: 2
Size: 15
Data Type: float64


Explanations:

* Shape: (5, 3) -> the dataset consists of 5 rows/samples (5 Experiments) & 3 features (Accuracy,
 Precision, Recall)

* No of Dimentions: 2 -> the dataset represents a 2D matrix (X,Y)

* Size: 15 -> The total measurments in the dataset = 15 measurement

* Data Type: float64 -> the dataset measurements are of type float (single-precision floating-point)

##### Q2

In [5]:
# Converting all metric values to percentages:
print(results * 100)

[[82. 78. 80.]
 [91. 88. 90.]
 [76. 81. 79.]
 [88. 84. 86.]
 [69. 72. 70.]]


### Part 2 — Select & Reshape

##### Q3

In [6]:
# (a) 
print(f"a. E2 Recall: {results[1][2]}")
# (b)
print(f"b. Accuracy: {results[:,0]}")
# (c)
print("c. Accuracy and Precision for the First Three Experiments:")
print(results[0:3,0:2])

a. E2 Recall: 0.9
b. Accuracy: [0.82 0.91 0.76 0.88 0.69]
c. Accuracy and Precision for the First Three Experiments:
[[0.82 0.78]
 [0.91 0.88]
 [0.76 0.81]]


##### Q4

In [7]:
arr = np.arange(24)

print(f"array: {arr}")
print(f"shape: {arr.shape}") # Vector of 24 Elements
print(f"size: {arr.size}")

array: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]
shape: (24,)
size: 24


In [8]:
# Reshaping
arr_reshaped = arr.reshape(4, 6)

print("reshaped array:")
print(arr_reshaped)
print(f"reshaped array shape: {arr_reshaped.shape}")
print(f"reshaped array size: {arr_reshaped.size}")


reshaped array:
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]
reshaped array shape: (4, 6)
reshaped array size: 24


In [9]:
# Flattening back
arr_flattened = arr_reshaped.ravel()
print(f"flattened array: {arr_flattened}")
print(f"flattened array shape: {arr_flattened.shape}")
print(f"flattened array size: {arr_flattened.size}")

flattened array: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]
flattened array shape: (24,)
flattened array size: 24


NOTE: since the 1D vector (24,) and 2D matrix (4, 6) share the same size = 24, so the reshaping is Valid

### Part 3 — Analyze the Results

##### Q5

In [10]:
# (a)
print(f"1. Overall Mean = {results.mean()}")
# (b)
print(f"2. Mean of each metric = {results.mean(axis=0)}")
# (c)
print(f"3. Mean score of each experiment = {results.mean(axis=1)}")

1. Overall Mean = 0.8093333333333332
2. Mean of each metric = [0.812 0.806 0.81 ]
3. Mean score of each experiment = [0.8        0.89666667 0.78666667 0.86       0.70333333]


Explanation:
1. Overall Mean = 0.81 -> "results.mean()" without specifying a certain axis renders the overall average of all matrix       
( average of all Experiment dataset measurements )

2. Mean of each metric = [0.812 0.806 0.81 ] -> "results.mean(axis=0)" renders the mean of each column in the matrix            
( average of each Metric "Accuracy, Precision, Recall")

3. Mean score of each experiment = [0.8 0.897 0.787 0.86 0.703] -> "results.mean(axis=1)" renders the mean of each row in the matrix            
( average of each Experiment "E1, E2, E3, E4, E5")

##### Q6

In [11]:
# (a)
print(f"1. Overall Standard Deviation = {results.std()}")
# (b)
print(f"2. Overall Variance = {results.var()}")

1. Overall Standard Deviation = 0.06826093725957444
2. Overall Variance = 0.004659555555555557


### Part 4 — Filter & Extend the Dataset

##### Q7

In [12]:
# Filtering
filtered_results = results[(results[:, 0] >= 0.80) & (results[:, 2] >= 0.80)]

print("Filtered Results:")
print(filtered_results)
print(f"filtered results has {filtered_results.shape[1]} colums")

Filtered Results:
[[0.82 0.78 0.8 ]
 [0.91 0.88 0.9 ]
 [0.88 0.84 0.86]]
filtered results has 3 colums


##### Q8

In [13]:
# Vertical Stacking
new_experiment = np.array([[0.85, 0.83, 0.87]])
print(f"New Experiment (E6): {new_experiment}")

results = np.vstack((results, new_experiment))
print("Updated Results with new Experiment (E6):")
print(results)

New Experiment (E6): [[0.85 0.83 0.87]]
Updated Results with new Experiment (E6):
[[0.82 0.78 0.8 ]
 [0.91 0.88 0.9 ]
 [0.76 0.81 0.79]
 [0.88 0.84 0.86]
 [0.69 0.72 0.7 ]
 [0.85 0.83 0.87]]


##### Q9

In [14]:
# Horizontal Stacking
f1 = np.zeros((results.shape[0], 1))
print("New Metric F1:")
print(f1)

results = np.hstack((results, f1))
print("Updated Results with new Metric (F1):")
print(results)

New Metric F1:
[[0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]]
Updated Results with new Metric (F1):
[[0.82 0.78 0.8  0.  ]
 [0.91 0.88 0.9  0.  ]
 [0.76 0.81 0.79 0.  ]
 [0.88 0.84 0.86 0.  ]
 [0.69 0.72 0.7  0.  ]
 [0.85 0.83 0.87 0.  ]]


In [15]:
# updating F1 
results[:, 3] = np.array([0.79, 0.89, 0.8, 0.85, 0.71, 0.86])

print("Updated (F1):")
print(results)

Updated (F1):
[[0.82 0.78 0.8  0.79]
 [0.91 0.88 0.9  0.89]
 [0.76 0.81 0.79 0.8 ]
 [0.88 0.84 0.86 0.85]
 [0.69 0.72 0.7  0.71]
 [0.85 0.83 0.87 0.86]]
